### PR TITLE
Add extension usage telemetry

### LINKED ISSUE #6109: Add extension usage telemetry
## Problem

k6 records which built-in modules a test uses but deliberately excludes extensions, so we have no visibility into which extensions people actually run. With binary provisioning now GA and a public catalog of official and community extensions, that blind spot blocks decisions about which extensions to support, promote, or retire. Cloud-executed runs are partly tracked, but local execution, which is most OSS usage, stays invisible.

## Significance

Extension adoption has become a core product signal: it tells us which community extensions deserve promotion to official, where to spend maintenance effort, and how well binary provisioning is being adopted versus custom builds. Today product, the k6 engine team, and extension maintainers all make those calls on guesswork. The catalog's value depends on knowing what its users actually reach for, and right now we are guessing.

## Cost of inaction

We keep promoting and deprecating extensions blind, spending effort on unused ones while missing popular community extensions that should become official. Every release we ship without this loses adoption history we can never recover. The decisions the catalog exists to support stay anecdote-driven.

## Desired state

Anyone can see which catalog extensions are used, at what version, and through which surface they were used (an imported module, an output, or a subcommand) for both local and cloud runs, regardless of whether the binary was built with xk6 or produced by binary provisioning. The data rides the existing anonymous usage report, so users keep one familiar opt-out and have nothing new to configure. Private and unlisted extensions are never reported, only the ones already public in the catalog, so the long-standing privacy concern that kept extensions out of telemetry stays addressed.

## Done when

- A used catalog importable extension (e.g. `k6/x/sql`) appears in the report with module, version, and `kind: js`.
- A used catalog subcommand (e.g. `k6 x docs`) is reported with `kind: subcommand`.
- A used catalog output extension (an `xk6-output-*` via `--out`) is reported with `kind: output`; built-in outputs like `json` are not. The public catalog doesn't list output extensions (no report until added to the catalog).
- Multiple extensions used in one run are each reported.
- Network or catalog failure never fails or delays the run: with a warm catalog cache a cataloged extension is still reported; with no cache, nothing is reported (fail closed).
- `--no-usage-report` / `K6_NO_USAGE_REPORT` suppresses all of it (run and subcommand paths); there is no separate opt-out.
- Built-in `modules` and `outputs` reporting is unchanged.
- A private, unlisted, or unknown extension is never reported, even when used.
- An extension baked into the binary but not used in the run is not reported.
- Reporting is identical for an xk6-built binary and a binary-provisioning-built binary.

## Out of scope

- Populating the catalog with output extensions.
- Counting unique users or installations ([telemetry issue](https://github.com/grafana/k6/issues/4077)).
- Telling apart xk6 builds from binary provisioning ([#6111](https://github.com/grafana/k6/issues/6111)).

## Related

- #6110 (spec and design)
- #6111
- #2952
- grafana/k6-cloud#3575

### PR DESCRIPTION
**Pull request by inancgumus**
## What?

Adds anonymous extension usage to the k6 usage report, covering the three user-facing surfaces:

- An extension (a `k6/x/` module) is reported when a run resolves it, not when it is merely compiled into the binary.
- An output extension is reported when a run selects it with `--out`; built-in outputs are unaffected.
- A subcommand extension is reported when invoked via `k6 x`, whether it exits zero or non-zero.

Each entry carries the extension's Go module path, version, and kind. Only extensions listed in the public registry catalog are reported: matching is by module path, so a private fork reusing a public import name is dropped, and an unreachable catalog reports nothing rather than falling back to unfiltered data. When no cataloged extension was used, the report omits the field entirely. The existing `--no-usage-report` / `K6_NO_USAGE_REPORT` opt-out gates all of it; no new opt-out, no new endpoint, and the send stays bounded and non-fatal: it never fails a command, and a hanging endpoint delays exit by at most 3 seconds, the same bound the `k6 run` report has always had.

## Why?

Extension adoption is a blind spot: the usage report tracks built-in modules but deliberately drops every extension, so decisions about which extensions to promote, maintain, or retire run on guesswork. Cloud-executed runs are partly tracked, while local execution, most OSS usage, stays invisible. Reporting catalog extensions through the same anonymous report k6 already sends closes that gap without a new data category: private and unlisted extensions never appear, xk6-built and provisioned binaries report identically, and a newly cataloged extension becomes reportable as soon as the catalog cache refreshes, with no new k6 release.

## Note

- #6110 holds the requirements, edge cases, and test vectors.
- The output surface ships now but reports nothing in production until the catalog lists output extensions.
- Distinguishing [xk6 builds](https://github.com/grafana/k6/pull/6111) from provisioned binaries is out of scope.
- One new env var, `K6_USAGE_REPORT_URL`, makes the report endpoint overridable so the integration tests can assert on the actually-sent report; catalog membership reuses the existing `K6_PROVISION_CATALOG_URL` override.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests for my changes.
- [x] I have run linter and tests locally (`make check`) and all pass.

## Related PR(s)/Issue(s)

- Closes #6109

### CODE REVIEW

**Review by szkiba — DISMISSED** [review]

File: internal/cmd/report.go [hunk 1]
```diff
@@ -7,40 +7,156 @@ import (
...
+		entries = append(entries, newExtensionEntry(e))
+	}
+	return entries
+}
+
+// reportUsage sends the report built by create, logging the attempt and outcome
+// at debug, all bounded by a timeout. Both k6 run and k6 x call it; whether to
+// run it in the background is the caller's concern, not the pipeline's.
+func reportUsage(ctx context.Context, gs *state.GlobalState, create func(ctx context.Context) map[string]any) {
+	reportCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
```
Comment (ankur22) [inline internal/cmd/report.go:135]: The 3 seconds used to be just for sending the report. Now the catalog fetch happens first inside the same window, so if the catalog is slow enough to use it all up, we lose the whole report — not just the extensions part. The tests cover a catalog that's down (fails instantly) but not one that's slow. Would it be worth giving the fetch its own smaller timeout so the report always has time to go out?

**Review by ankur22 — COMMENTED** [review]
Really nice work!

A few questions inline, none blocking. Two small things here:

- The description says a slow endpoint "never delays or fails a command", but `k6 x` sends its report before exiting, so a hanging endpoint can add up to 3 seconds to the exit. Probably fine, just worth softening the wording.
- The usage-collection docs will need an update — `k6 x` now sends a report, and the run report has a new `extensions` field. Is that tracked somewhere?

File: internal/cmd/run.go [hunk 2]
```diff
@@ -460,15 +460,9 @@ func (c *cmdRun) run(cmd *cobra.Command, args []string) (err error) {
...
-			defer reportCancel()
-			logger.Debug("Sending usage report...")
-
-			if rerr := reportUsage(reportCtx, execScheduler, test); rerr != nil {
-				logger.WithError(rerr).Debug("Error sending usage report")
-			} else {
-				logger.Debug("Usage report sent successfully")
-			}
+			reportUsage(globalCtx, c.gs, func(ctx context.Context) map[string]any {
+				return createReport(test.preInitState.Usage, execScheduler, catalogModulePaths(ctx, c.gs))
```
Comment (ankur22) [inline internal/cmd/run.go:464]: This fetches the catalog on every run where the cache is missing or stale, even when the script didn't use any extensions — which is most runs. It's cheap and cached for a day, so not a big deal, but did you consider skipping the fetch when there's nothing to filter?

File: internal/cmd/subcommand.go [hunk 3]
```diff
@@ -83,6 +83,27 @@ Run "k6 x explore" to see the full list of official and community-provided subco
...
+// reportSubcommandUsage reports usage for a subcommand extension after it runs.
+// It is called with the command cobra actually executed, whether that command
+// succeeded or failed, so a subcommand that exits non-zero is still counted. It
+// reports only for a name backed by a registered subcommand extension, so the
+// catch-all `x` help, `k6 run`, completions, and provisioning stubs are ignored.
+func reportSubcommandUsage(gs *state.GlobalState, executed *cobra.Command) {
+	if executed == nil {
+		return
+	}
+	extension, ok := ext.Get(ext.SubcommandExtension)[executed.Name()]
```
Comment (ankur22) [inline internal/cmd/subcommand.go:95]: Matching just on the command name has two edge cases: an extension named the same as a builtin (`status`, `version`, …) would get reported when someone runs the builtin, and an extension with nested subcommands wouldn't get reported at all, since cobra returns the child command. Both unlikely today, but checking that the command actually lives under `x` would fix both. Worth doing now, or a follow-up?

File: internal/cmd/subcommand.go [hunk 4]
```diff
@@ -83,6 +83,27 @@ Run "k6 x explore" to see the full list of official and community-provided subco
...
+	if !ok {
+		return
+	}
+	if conf, err := readEnvConfig(gs.Env); err != nil || conf.NoUsageReport.Bool {
```
Comment (ankur22) [inline internal/cmd/subcommand.go:99]: Small gap: setting `noUsageReport: true` in the config file turns off the run report but not this one, since only env vars are checked here. Was that deliberate, to avoid loading the config file for every command? If so, probably worth a docs note that the env var is the one that covers everything.

File: internal/cmd/report.go (hunk 1 above)
Comment (inancgumus) [reply internal/cmd/report.go:135]: TBH, I considered this too, but a smaller timeout wouldn't help much. It'd still eat from the same 3s, so the `POST` doesn't get its old budget back. We'd just lose the extensions data instead of the whole report. Also, most runs don't fetch at all, as the cache stays fresh for 24h (plus: `k6 x` warms the same cache too), so the `POST` keeps the whole 3s. I'd like to leave it as is until it becomes a real problem.

File: internal/cmd/run.go (hunk 2 above)
Comment (inancgumus) [reply internal/cmd/run.go:464]: It'd be a nice optimization, so I'd rather do it in a follow-up PR.

File: internal/cmd/subcommand.go (hunk 3 above)
Comment (inancgumus) [reply internal/cmd/subcommand.go:95]: Good catch! Let's fix it in the follow-up stacked PR 🤝

File: internal/cmd/subcommand.go (hunk 4 above)
Comment (inancgumus) [reply internal/cmd/subcommand.go:99]: Another good one! I'll fix it in the follow-up stacked PR as well.

**Comment by inancgumus** [comment]
Great review, @ankur22! Yes, [Task #19](https://github.com/grafana/k6/pull/6110/changes#diff-4ab8c3d53495c8c617df7454590541606609bdcde8897c219dfd8a5150d246e4R144-R148) tracks the docs update, which I'll do once this PR merges 🤝 I've also reworded the PR description. For the review requests, I've created [a stacked follow-up PR](https://github.com/grafana/k6/pull/6224), which we can merge after this one.

**Review by ankur22 — APPROVED** [review]
LGTM 🚀

File: internal/cmd/report.go [hunk 5]
```diff
@@ -7,40 +7,156 @@ import (
...
 
-func createReport(u *usage.Usage, execScheduler *execution.Scheduler) map[string]any {
-	execState := execScheduler.GetState()
-	m := u.Map()
+// extensionEntry identifies a used extension in the usage report by its Go
+// module path, along with its version and kind.
+type extensionEntry struct {
+	Module  string `json:"module"`
+	Version string `json:"version"`
+	Kind    string `json:"kind"`
```
Comment (janHildebrandt98) [inline internal/cmd/report.go:24]: Why not keep this as "Type" so we don't introduce another term for the same thing?

**Review by janHildebrandt98 — APPROVED** [review]

File: internal/cmd/report.go (hunk 5 above)
Comment (inancgumus) [reply internal/cmd/report.go:24]: I'll sort it out in the next PR 🤝

Comment (inancgumus) [reply internal/cmd/report.go:24]: https://github.com/grafana/k6/pull/6229
